lib/http.c

```diff
@@ -1456,8 +1456,10 @@ CURLcode Curl_http_done(struct connectdata *conn,
      data->state.negotiate.state == GSS_AUTHSENT) {
     /* add forbid re-use if http-code != 401/407 as a WA only needed for
      * 401/407 that signal auth failure (empty) otherwise state will be RECV
-     * with current code */
-    if((data->req.httpcode != 401) && (data->req.httpcode != 407))
+     * with current code.
+     * Do NOT close successful CONNECT-only connections. */
+    if((data->req.httpcode != 401) && (data->req.httpcode != 407) &&
+       (!data->set.connect_only || data->req.httpcode != 200))
```

Comment:
Why do you need the '|| data->req.httpcode != 200' part?
It defeats the purpose of this whole section.
Does it work for you if you omit it?

In my opinion, we could indeed escape the check in 'connect_only' mode as libcurl has no longer control over the connection anyway so its the responsibility of the application to use the authenticated connection correctly.